### ISSUE TITLE
Android:FlutterView.getBitmap can't capture Image&Videos.

### ISSUE THREAD

**Issue by kangwang1988** [flutter/flutter#16412]
## Steps to Reproduce

Clone code at: git@github.com:FlutterRepo/flutter_app.git

Click "Snapshot" Button in the AppBar

FlutterView.getBitmap will be called and an Image will be saved.However, the image seems to be missing images and videos.

App:
![screenshot_20180410-170551](https://user-images.githubusercontent.com/817851/38547690-52a6bf3e-3ce2-11e8-9a2f-7b9f3a593bd4.jpg)
Snapshot:
![snapshot](https://user-images.githubusercontent.com/817851/38547699-58c964f2-3ce2-11e8-80be-45eb22c78f49.jpg)

Please tell us what you were doing and what went wrong. If you are running flutter tools from the command line, please try adding the `-v` or `--verbose` option to gather more information.

If the problem is with your application's rendering, please attach a screenshot and any relevant source code.
If you are getting an exception in the logs, and your code is implicated in the first few frames, then please include the source code for the functions involved.

## Logs

Run your application with `flutter run` and attach all the log output.

Run `flutter analyze` and attach any output of that command also.

## Flutter Doctor

Paste the output of running `flutter doctor -v` here.

KyleWongdeMacBook-Pro:flutter_app kylewong$  flutter doctor -v
[✓] Flutter (Channel beta, v0.2.8, on Mac OS X 10.13.4 17E199, locale en-CN)
    • Flutter version 0.2.8 at /Users/kylewong/Downloads/flutter
    • Framework revision b397406561 (8 days ago), 2018-04-02 13:53:20 -0700
    • Engine revision c903c217a1
    • Dart version 2.0.0-dev.43.0.flutter-52afcba357

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/kylewong/Library/Android/sdk
    • Android NDK at /Users/kylewong/Library/Android/sdk/ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.3, Build version 9E145
    • ios-deploy 1.9.2
    • CocoaPods version 1.2.0

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] Connected devices (1 available)
    • MHA AL00 • GWY7N16B02004245 • android-arm64 • Android 8.0.0 (API 26)

• No issues found!

> For more information about diagnosing and reporting Flutter bugs, please see [https://flutter.io/bug-reports/](https://flutter.io/bug-reports/).

**Comment by kangwang1988** [comment]
@eseidelGoogle 
@korbin-w

**Comment by brianosman** [comment]
Status: I've managed to repro this issue (on a Nexus 6P). Still need to debug it, but I'm guessing that I'll have an answer tomorrow. I suspect it's not specific to any GPUs.

**Comment by kangwang1988** [comment]
@eseidelGoogle 
Yes, it's a blocker. As we realized a hybrid page stack(native and flutter) where there is only one flutter context. We need to save the snapshot so that, the push/pop animation can looks fine. 
Currently, the hybrid stack works fine with only one flutter instance except the getbitmap function.

**Comment by sigurdm** [comment]
I agree this is a bug in the engine, and not Skia's fault.

In `getBitMap()` in `platform_view_android.cc` we create an `SkSurface::MakeRasterDirect`. It doesn't get connected to the skia GrContext that holds images and textures (and the raster cache). Therefore the images and videos do not show up.

I think we need to use `SkSurface::makeSurface( const SkImageInfo & )` to create a surface with the same GrContext and copy the bitmap after drawing. This could be handled by the Rasterizer - we have at least two more places doing screenshotting in the engine(`FlutterView.mm`, and `platform_view_service_protocol.cc`) these call paths also seem to suffer from https://github.com/flutter/flutter/issues/16143.

cc @cbracken.

**Comment by brianosman** [comment]
FYI: This was my conclusion as well - and I was going to suggest that the engine use a GPU surface (rather than a raster surface).

**Comment by chinmaygarde** [comment]
https://github.com/flutter/engine/pull/4979 fixes snapshotting of scenes with images uploaded on the IO context. That patch is enough to fix this issue.

There is a separate issue where external textures are not snapshotted. This is harder to support because the the engine does not own those textures.

Upon reading [this comment](https://github.com/flutter/flutter/issues/16412#issuecomment-380297178), it does not look like @kangwang1988 will be running into this edge case. Can you confirm? In any case, we can close this issue once the patch lands.

**Comment by kangwang1988** [comment]
@chinmaygarde 
In our project, the snapshot for iOS is not a problem.
The problem occurs in Android. The FlutterView.getBitmap fails when there is a Image/Video.

**Comment by kangwang1988** [comment]
KyleWongdeMacBook-Pro:flutter kylewong$ /Users/kylewong/Downloads/flutter/bin/flutter doctor -v
[✓] Flutter (Channel master, v0.3.1-pre.15, on Mac OS X 10.13.4 17E199, locale en-CN)
    • Flutter version 0.3.1-pre.15 at /Users/kylewong/Downloads/flutter
    • Framework revision f049f436c8 (6 hours ago), 2018-04-16 14:17:55 -0700
    • Engine revision 76cb311d9c
    • Dart version 2.0.0-dev.47.0.flutter-f76dad0adc

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/kylewong/Library/Android/sdk
    • Android NDK at /Users/kylewong/Library/Android/sdk/ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.3, Build version 9E145
    • ios-deploy 1.9.2
    • CocoaPods version 1.2.0

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 23.2.2
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] Connected devices (2 available)
    • MHA AL00          • GWY7N16B02004245                         • android-arm64 • Android 8.0.0 (API 26)
    • KyleWong's iPhone • 1c8e085cf2ff6fa27643ab4afec4bf4a077688af • ios           • iOS 11.3

• No issues found!


I tried with this master branch, however, the FlutterView.getBitmap for Android still fails.

**Comment by kangwang1988** [comment]
Update:
FlutterView.getBitmap may cause memory leak, and result in OOM.
We found dozens of crash with stacktrace saying:
crash堆栈：
    00 pc 0004a720  /system/lib/libc.so (tgkill+12)
    01 pc 00047eb3  /system/lib/libc.so (pthread_kill+34)
    02 pc 0001d955  /system/lib/libc.so (raise+10)
    03 pc 000194a1  /system/lib/libc.so (__libc_android_abort+34)
    04 pc 000170e4  /system/lib/libc.so (abort+4)
    05 pc 0031059d  /system/lib/libart.so (art::Runtime::Abort()+252)
    06 pc 000a9f77  /system/lib/libart.so (art::LogMessage::~LogMessage()+866)
    07 pc 00329f7f  /system/lib/libart.so (art::Thread::AssertNoPendingException() const+602)
    08 pc 00181219  /system/lib/libart.so (art::gc::Heap::AllocateInternalWithGc(art::Thread*, art::gc::AllocatorType, bool, unsigned int, unsigned int*, unsigned int*, unsigned int*, art::mirror::Class**)+32)
    09 pc 000f5b6f  /system/lib/libart.so (art::mirror::Object* art::gc::Heap::AllocLargeObject(art::Thread*, art::mirror::Class**, unsigned int, art::mirror::SetLengthVisitor const&)+706)
    10 pc 000f4c61  /system/lib/libart.so (art::mirror::Object* art::gc::Heap::AllocObjectWithAllocator(art::Thread*, art::mirror::Class*, unsigned int, art::gc::AllocatorType, art::mirror::SetLengthVisitor const&)+72)
    11 pc 000d6cd9  /system/lib/libart.so (art::mirror::PrimitiveArray::Alloc(art::Thread*, unsigned int)+72)
    12 pc 00276015  /system/lib/libart.so (_jintArray* art::JNI::NewPrimitiveArray<_jintArray*, art::mirror::PrimitiveArray >(_JNIEnv*, int)+412)
    13 pc 00050fe3  /data/app/com.fish.fish-1/lib/arm/libflutter.so
    14 pc 00052aff  /data/app/com.fish.fish-1/lib/arm/libflutter.so
    15 pc 000599e1  /data/app/com.fish.fish-1/lib/arm/libflutter.so
    16 pc 0005c223  /data/app/com.fish.fish-1/lib/arm/libflutter.so
    17 pc 00011ee3  /vendor/lib/libutils.so (android::Looper::pollInner(int)+614)
    18 pc 00011bef  /vendor/lib/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+26)
    19 pc 0000a683  /system/lib/libandroid.so (ALooper_pollOnce+50)
    20 pc 0005c1df  /data/app/com.fish.fish-1/lib/arm/libflutter.so
    21 pc 00059a6d  /data/app/com.fish.fish-1/lib/arm/libflutter.so
    22 pc 0005ae15  /data/app/com.fish.fish-1/lib/arm/libflutter.so
    23 pc 00047983  /system/lib/libc.so (__pthread_start(void*)+22)
    24 pc 00019efd  /system/lib/libc.so (__start_thread+6)
--------------------------------------------------------------------------------------------------------------------------------------------------
可疑日志：
    W  art   Throwing OutOfMemoryError "Failed to allocate a 13916172 byte allocation with 6929272 free bytes and 6MB until OOM"
    F  art   art/runtime/thread.cc:1663] No pending exception expected: java.lang.OutOfMemoryError: Failed to allocate a 13916172 byte allocation with 6929272 free bytes and 6MB until OOM
    F  art   art/runtime/thread.cc:1663] (Throwable with empty stack trace)
    F  art   art/runtime/runtime.cc:408] Runtime aborting...

When I comment out all call for FlutterView.getBitmap, it works fine. But when I use it,even when I just call it and don't use the result Bitmap in android to render, this OOM error will appear. So I doubt that call for this method could course memory leak. 
@chinmaygarde 
@eseidelGoogle

**Comment by eseidelGoogle** [comment]
@jason-simmons may also know.

**Comment by jason-simmons** [comment]
Do you have this patch?
    https://github.com/flutter/engine/commit/adee351424f6b4d11b9ee87699dc5edf2d7f1e48

With that patch, the call to JNI NewIntArray in GetBitmap should return to the Java caller and throw an OutOfMemoryError if the NewIntArray allocation fails.

**Comment by kangwang1988** [comment]
@jason-simmons 
I've tried to apply that patch. However, we're at beta branch(engine version: e61bb9ac3a3fd789754e2e54220bcfc27076a857 in March 14th) and still not the latest beta one.
So I can't merge your commit and made some modifications by myself.
As below:

<img width="1120" alt="screen shot 2018-04-24 at 10 51 46 am" src="https://user-images.githubusercontent.com/817851/39163681-1c236fda-47ae-11e8-8800-d41b978542b9.png">


However,after I made some modifications, the abort disappears. However, the OOM appears in many devices. After I remove the getBitmap call, it seems fine.

So I've already remove the getBitmap call temporarily.
After the getBitmap function can work fine, we'll apply the method again.

**Comment by Hixie** [comment]
@korbin-w Very well! Keep us posted!

**Comment by kangwang1988** [comment]
We've upgraded our flutter to betav0.3.1.
I'll check this issue again.

**Comment by kangwang1988** [comment]
@chinmaygarde @Hixie 
I tested it just now, the Flutter.getBitmap works fine for image.
However, it fails for video. The video area is blank in our snapshot.
Can you guys have a look at it?
The video is implemented with a texture approach.

**Comment by andreidiaconu** [comment]
@chinmaygarde @Hixie
Is this possibly related? https://github.com/flutter/flutter/issues/17687
I'm trying to figure this out for a few days and is a major blocker for me. I can give more details if needed.
